### PR TITLE
Undo struct promotion on "RetInd" code path

### LINKED ISSUE #57912: [JitStressWithRandomNumbers] seq_short_1_d fails
Example:
https://helixre8s23ayyeko0k025g8.blob.core.windows.net/dotnet-runtime-refs-pull-57900-merge-f26c04388af241ca9d/JIT.Methodical.eE/1/console.0f01af6b.log?sv=2019-07-07&se=2021-09-12T00%3A10%3A43Z&sr=c&sp=rl&sig=xX%2FwSE9hoH%2FWaOhtQFIwGMXAc9V3kIgeQes1vC4jiNM%3D

Repro steps (windows x64):
```
set COMPlus_TieredCompilation=0
set COMPlus_JitStress=99
JIT\Methodical\explicit\coverage\seq_short_1_d\seq_short_1_d.cmd
```

Error log:
```
      Expected: 100
      Actual: 118
      END EXECUTION - FAILED
```
the issue from a quality week JitStress experiment.

### PR DESCRIPTION
**Pull request by kunalspathak**
When we undo the struct promotion, we revert all the promoted fields except in the code path where we would return the promoted field. We were keeping the promoted field leading to bad codegen.

No asmdiff on libraries/benchmarks/coreclr_test. 

Since the original repro needed jitstress switch, I verified that this PR fixes the code we geenrate.

Fixes: https://github.com/dotnet/runtime/issues/57912

### CODE REVIEW

**Comment by kunalspathak** [comment]
@dotnet/jit-contrib

**Review by AndyAyersMS — APPROVED** [review]
Changes look good. I assume `fgMorphRetInd` is relatively new?

Is there any way to add checking for cases where we mess up this undo?

**Comment by kunalspathak** [comment]
> Changes look good. I assume `fgMorphRetInd` is relatively new?

It was added last year in https://github.com/dotnet/runtime/pull/37745

> Is there any way to add checking for cases where we mess up this undo?

I tried looking around and I didn't see a good place to have this check. What I ended up doing is introducing a field `lvUndoneStructPromotion` and then, during mark local reference, whenever we see that we are increasing the reference count of a struct field, we make sure that its parent struct didn't have `lvUndoneStructPromotion` set.

By running superpmi replay on coreclr, it hits the newly added assert on exactly same tests.

File: src/tests/JIT/Regression/JitBlue/Runtime_57912/Runtime_57912.csproj [hunk 1]
```diff
@@ -0,0 +1,20 @@
...
+    <Optimize>True</Optimize>
+    <DebugType>None</DebugType>
+  </PropertyGroup>
+  <ItemGroup>
+    <Compile Include="$(MSBuildProjectName).cs" />
+  </ItemGroup>
+  <PropertyGroup>
+    <CLRTestBatchPreCommands><![CDATA[
+$(CLRTestBatchPreCommands)
+set DOTNET_JitEnableFinallyCloning=0
```
Comment (AndyAyersMS) [inline src/tests/JIT/Regression/JitBlue/Runtime_57912/Runtime_57912.csproj:13]: Is this really needed to repro the failure? I don't see any finallies in the test case.

Comment (kunalspathak) [reply src/tests/JIT/Regression/JitBlue/Runtime_57912/Runtime_57912.csproj:13]: ah..it was a copy paste from different project.

**Review by sandreenko — APPROVED** [review]

**Pull request merged**